### ISSUE TITLE
The Module API docs' DemoResource crashes

### ISSUE THREAD

**Issue by cvwright** [matrix-org/synapse#10837]
### Description

The Module API docs include an example module, which is extremely helpful for a new module developer.

However, the DemoResource's `render_GET` function causes a `TypeError` during JSON encoding and fails to return the intended response.

### Steps to reproduce

- Install the DemoModule.
  - I put the example code in `synapse/demo_module/__init__.py`
  - Copy the example config verbatim into `homeserver.yaml`
- Start Synapse: 
```console
$ python3 -m synapse.app.homeserver --config-path homeserver.yaml
```
- Load the demo URL `_synapse/client/demo/hello?name=world` in a browser or use cURL/wget/etc

- **Homeserver**: My own development server

If not matrix.org:

- **Version**: 1.42.0

- **Install method**: Git clone

- **Platform**: Debian 10, x86, Python 3.7.3

### Proposed Fix

Since the problem seems to be with the JSON encoder failing to handle bytes, we can simply decode the string before we put it into the JSON.  Also, it seems that we need to encode the JSON string back to bytes before we return.

```python
    def render_GET(self, request: Request):
        name = request.args.get(b"name")[0].decode() 
        request.setHeader(b"Content-Type", b"application/json") 
        return json.dumps({"hello": name}).encode()
```

**Comment by DMRobertson** [comment]
Thanks for spotting! Would you be up for making this change into a pull request?

I'm guessing the example you mention is [the one here](https://matrix-org.github.io/synapse/latest/modules.html#example)?  I wonder if that was written a while back with Python 2 or a particularly old twisted version in mind.

> Also, it seems that we need to encode the JSON string back to bytes before we return.

Indeed: twisted expects `render_METHOD` to return a `bytes` type if it doesn't write the result asynchronously. See also https://twistedmatrix.com/documents/21.2.0/api/twisted.web.resource.Resource.html#render

(cc @babolivier  because modules.)

**Comment by cvwright** [comment]
Glad to be able to help!  And yes, that's the example that I was looking at.

> Would you be up for making this change into a pull request?

Sure thing.  I'll try to submit the PR tomorrow if I can find some time.

Thanks for the quick response, and for confirming that it's not just something weird I'm doing on my end.

I'm loving the Module API so far btw.

**Comment by cvwright** [comment]
So it turns out that the documentation I was reading had already been updated to be even more detailed, and there's no more `DemoModule` or `DemoResource`.  But it appears that the same issue exists in the new `IsUserEvilResource`.  So the linked PR fixes it there.

**Comment by richvdh** [comment]
fixed by #10845. @anoadragon453 please can you make sure that PRs and issues are correctly linked before merging.

**Comment by anoadragon453** [comment]
@richvdh Oops, yep can do.
